src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,462 @@
+
+            if (shouldCompile)
+            {
+                DesignTimeInputsItem inputs = AppliedValue.Value;
+                HashSet<string> files = GetFilesToCompile(fileName, inputs.SharedInputs);
+                string outputFileName = GetOutputFileName(inputs.OutputPath, fileName);
+                await CompileTempPEAsync(files, outputFileName);
+            }
+
+            _buildManager.OnDesignTimeOutputDirty(fileName);
+        }
+
+        public async Task<string> GetTempPEDescriptionXmlAsync(string fileName)
+        {
+            await InitializeAsync();
+
+            await _unconfiguredProjectServices.ThreadingService.SwitchToUIThread();
+
+            DesignTimeInputsItem inputs = AppliedValue.Value;
+            if (fileName == null)
+                throw new ArgumentException("Must supply a file to build", nameof(fileName));
+
+            if (!inputs.Inputs.Contains(fileName))
+                throw new ArgumentException("FileName supplied must be one of the DesignTime source files", nameof(fileName));
+
+            string outputPath = inputs.OutputPath;
+
+            HashSet<string> files = GetFilesToCompile(fileName, inputs.SharedInputs);
+            string outputFileName = GetOutputFileName(outputPath, fileName);
+            if (CompilationNeeded(files, outputFileName))
+            {
+                // For parity with legacy we don't care about the compilation result: Legacy only errors here if it runs out of memory queuing the compilation
+                await CompileTempPEAsync(files, outputFileName);
+            }
+
+            // VSTypeResolutionService is the only consumer, and it only uses the codebase element so it's fine to default most of these (VC++ does the same)
+            return $@"<root>
+  <Application private_binpath = ""{outputPath}""/>
+  <Assembly
+    codebase = ""{outputFileName}""
+    name = ""{fileName}""
+    version = ""0.0.0.0""
+    snapshot_id = ""1""
+    replaceable = ""True""
+  />
+</root>";
+        }
+
+        internal bool CompilationNeeded(HashSet<string> files, string outputFileName)
+        {
+            if (!_fileSystem.FileExists(outputFileName))
+                return true;
+
+            DateTime outputDateTime = _fileSystem.LastFileWriteTime(outputFileName);
+
+            foreach (string file in files)
+            {
+                DateTime fileDateTime = _fileSystem.LastFileWriteTime(file);
+                if (fileDateTime > outputDateTime)
+                    return true;
+            }
+
+            return false;
+        }
+
+        private static string GetOutputFileName(string outputPath, string fileName)
+        {
+            return Path.Combine(outputPath, fileName.Replace('\\', '.') + ".dll");
+        }
+
+        protected virtual Task CompileTempPEAsync(HashSet<string> filesToCompile, string outputFileName)
+        {
+            return _sequentialTaskQueue.ExecuteTask(async () =>
```

Comment:
The sequential task queue is here to ensure that TempPE files get compiled in the order they were asked for. I cannot fathom a way that there could ever be multiple tasks in the queue because we always block the UI thread when compiling, but I left this here for parity with legacy, and to trigger some discussion around it.

Legacy has an event queue that it adds events to in order to compile TempPEs, however it too waits on the UI thread for a compilation, and it never adds an item to the queue without also immediately draining the queue and compiling all items, so I similarly cannot fathom a way that legacy could ever has multiple items in its queue. If anyone has any context around that, or wants to go and look at [the code](https://devdiv.visualstudio.com/DevDiv/_git/VS?path=%2Fsrc%2Fvsproject%2Flangbuild%2FCLangTempPEManager.cpp&version=GBmaster&line=795&lineStyle=plain&lineEnd=796&lineStartColumn=1&lineEndColumn=1) and work out why it might have a queue, I'd welcome it.

If nobody can work out why I don't see a reason for this to exist, though it is a fairly low overhead insurance policy.